staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go

```diff
@@ -275,33 +281,58 @@ func (evtv eachValTagValidator) getListValidations(fldPath *field.Path, t *types
 	result := Validations{}
 	result.OpaqueValType = validations.OpaqueType
 
-	var listMap *listMap
+	var listMetadata *listMetadata
 	if lm, found := evtv.byFieldPath[fldPath.String()]; found {
-		if !lm.declaredAsMap {
-			return Validations{}, fmt.Errorf("found listMapKey without listType=map")
+		if lm.declaredAsSet && lm.declaredAsMap {
+			return Validations{}, fmt.Errorf("listType cannot be both et and map")
 		}
-		if len(lm.keyFields) == 0 {
+		if lm.declaredAsMap && len(lm.keyFields) == 0 {
 			return Validations{}, fmt.Errorf("found listType=map without listMapKey")
 		}
-		listMap = lm
+		if len(lm.keyFields) > 0 && !lm.declaredAsMap {
+			return Validations{}, fmt.Errorf("found listMapKey without listType=map")
+		}
+		listMetadata = lm
 	}
+
 	for _, vfn := range validations.Functions {
 		var cmpArg any = Literal("nil")
-		if listMap != nil {
-			cmpFn := FunctionLiteral{
-				Parameters: []ParamResult{{"a", t.Elem}, {"b", t.Elem}},
-				Results:    []ParamResult{{"", types.Bool}},
-			}
-			buf := strings.Builder{}
-			buf.WriteString("return ")
-			for i, fld := range listMap.keyFields {
-				if i > 0 {
-					buf.WriteString(" && ")
+		if listMetadata != nil {
+			if listMetadata.declaredAsMap {
+				cmpFn := FunctionLiteral{
+					Parameters: []ParamResult{{"a", t.Elem}, {"b", t.Elem}},
+					Results:    []ParamResult{{"", types.Bool}},
+				}
+				buf := strings.Builder{}
+				buf.WriteString("return ")
+				for i, fld := range listMetadata.keyFields {
+					if i > 0 {
+						buf.WriteString(" && ")
+					}
+					buf.WriteString(fmt.Sprintf("a.%s == b.%s", fld, fld))
+				}
+				cmpFn.Body = buf.String()
+				cmpArg = cmpFn
+			} else if listMetadata.declaredAsSet {
+				elemType := t.Elem
+				if elemType.Kind == types.Alias {
```

Comment:
I think you want `realType()` here - pointers are comparable, but that's not what we want.  And a comment.